File: apps/api/internal/provider/factory.go
package provider

import (
	"context"
	"fmt"

	"github.com/manpreet/chatbot/apps/api/internal/config"
	"github.com/manpreet/chatbot/apps/api/internal/store"
)

func NewChatClient(cfg config.ChatConfig) ChatCompleter {
	switch cfg.Provider {
	case "openrouter":
		return NewOpenRouterClient(cfg)
	case "huggingface":
		return NewHuggingFaceClient(cfg)
	default:
		return unsupportedClient{provider: cfg.Provider}
	}
}

type unsupportedClient struct {
	provider string
}

func (c unsupportedClient) Complete(_ context.Context, _ []store.Message, _ int) (ChatResult, error) {
	return ChatResult{}, fmt.Errorf("unsupported chat provider %q", c.provider)
}

File: apps/pyapi/pyapi/providers/factory.py
from __future__ import annotations

from pyapi.config import ChatConfig

from .huggingface import HuggingFaceProvider
from .openrouter import OpenRouterProvider
from .types import ChatProvider, ChatResult


def create_chat_provider(config: ChatConfig) -> ChatProvider:
    if config.provider == "openrouter":
        return OpenRouterProvider(config)
    if config.provider == "huggingface":
        return HuggingFaceProvider(config)
    return UnsupportedProvider(config.provider)


class UnsupportedProvider:
    def __init__(self, provider: str):
        self.provider = provider
        self.model = ""

    async def complete(self, *_object: object) -> ChatResult:
        raise ValueError(f'unsupported chat provider "{self.provider}"')
